lib/awful/popup.lua

```diff
@@ -0,0 +1,434 @@
+---------------------------------------------------------------------------
+--- An auto-resized, free floating or modal wibox built around a widget.
+--
+-- This type of widget box (wibox) is auto closed when being clicked on and is
+-- automatically resized to the size of its main widget.
+--
+-- Note that the widget itself should have a finite size. If something like a
+-- `wibox.layout.flex` is used, then the size would be unlimited and an error
+-- will be printed. The `wibox.layout.fixed`, `wibox.container.constraint`,
+-- `forced_width` and `forced_height` are recommended.
+--
+--@DOC_awful_popup_simple_EXAMPLE@
+--
+-- Here is an example of how to create an alt-tab like dialog by leveraging
+-- the `awful.widget.tasklist`.
+--
+--@DOC_awful_popup_alttab_EXAMPLE@
+--
+-- @author Emmanuel Lepage Vallee
+-- @copyright 2016 Emmanuel Lepage Vallee
+-- @classmod awful.popup
+---------------------------------------------------------------------------
+local wibox     = require( "wibox"           )
+local util      = require( "awful.util"      )
+local glib      = require( "lgi"             ).GLib
+local placement = require( "awful.placement" )
+
+local module = {}
+
+local main_widget = {}
+
+-- Get the optimal direction for the wibox
+-- This (try to) avoid going offscreen
+local function set_position(self)
+    local pf = self._private.placement
+
+    if pf == false then return end
+
+    if pf then
+        pf(self, {bounding_rect = self.screen.geometry})
+        return
+    end
+
+    local geo = self._private.widget_geo
+
+    if not geo then return end
+
+    local preferred_positions = self._private.preferred_directions or {
+        "right", "left", "top", "bottom"
+    }
+
+    local preferred_anchors = self._private.preferred_anchors or {
+        "back", "front", "middle"
+    }
+
+    local _, pos_name, anchor_name = placement.next_to(self, {
+        preferred_positions = preferred_positions,
+        geometry            = geo,
+        preferred_anchors   = preferred_anchors,
+        offset              = self._private.offset or { x = 0, y = 0},
+    })
+
+    if pos_name ~= self._private.current_position then
+        self:emit_signal("property::current_position", pos_name)
+        self._private.current_position = pos_name
+    end
+
+    if anchor_name ~= self._private.current_anchor then
+        self:emit_signal("property::current_anchor", anchor_name)
+        self._private.current_anchor = anchor_name
+    end
+end
+
+-- Fit this widget into the given area
+function main_widget:fit(context, width, height)
+    if not self.widget then
+        return 0, 0
+    end
+
+    return wibox.widget.base.fit_widget(self, context, self.widget, width, height)
+end
+
+-- Layout this widget
+function main_widget:layout(context, width, height)
+    if self.widget then
+        local w, h = wibox.widget.base.fit_widget(self, context, self.widget, 9999, 9999)
```

Comment:
> but I feel like they should do so way "harder"

What do you have in mind? I don't feel like adding memes to the doc ;)

> How about using context.screen.geometry.width/height here instead?

I could add a new `awful.screen` function to get the maximum width/height of the screens. I can't just use the current screen because it can change in some async events.